### PR TITLE
[AVRO-2806] Upgrade from Netty 3 -> Netty 4

### PR DESCRIPTION
**Pull request by dkulp**
Updates Netty from 3.x to the latest 4.x release.  

This is a binary incompatible change.   Users of netty-rpc that have configured custom socket factories or thread pools and similar will need to migrate to the Netty 4 versions of those concepts.

### CODE REVIEW

**Comment by zeshuai007** [comment]
Hi @dkulp ,

Thanks to dkulp for solving this issue, it looks good.
Before the last week, I almost finished the netty3-> 4 API upgrade and refactoring locally. At that time, my headache was the change of the netty3-4 thread model, which caused TestNetty4XServerConcurrentExecution to fail. Unfortunately, I have something else Blocked my refactoring process.
So please forgive my recklessness, I hope to make a record here:
1. Changed the package path in netty4.X, there is no jboss, because the author left the jboss company
2. SimpleChannelUpstreamHandler-> SimpleChannelInboundHandler, deprecated channelOpen, use channelActive
3. FrameDecoder-> ByteToMessageDecoder
4. OneToOneEncoder-> MessageToMessageEncoder
5, write-> writeAndFlush, netty4 separates write and flush, write is not processing flush
6. NioServerSocketChannelFactory and NioClientSocketChannelFactory have been deprecated in netty 4.x. If there are no more complicated requirements in Netty 4.x, it is strongly recommended to use ServerBootstrap.channel (NioServerSocketChannel.class)
7, netty4.x changed the thread model, Handler registered in ChannelPipeline is serial. The related processing of serialization should use the new NioEventLoop, isolated from the i / o thread, otherwise there will be unexpected results in some cases, such as thread blocking
(The above is purely personal understanding, for reference only!)

What I am still confused about is:
addLast (new DefaultEventLoopGroup (n), new NettyServerAvroHandler ()) does not seem to work.
Start a new thread in channelRead0
TestNetty4XServerConcurrentExecution test passed

**Comment by zeshuai007** [comment]
NettyServer  -> new DefaultEventLoopGroup (16);
I am not sure whether the number of threads 16 meets the needs of the job.
In practice and application, regardless of memory consumption, I like to use the thread number formula to calculate
Nthreads = Ncpu * Ucpu * (1 + w / c);
So I think 16 can be used as the default value and the settings can be passed in through the constructor.

File: lang/java/ipc-netty/src/main/java/org/apache/avro/ipc/netty/NettyServer.java [hunk 1]
```diff
@@ -59,55 +55,32 @@
...
+        .channel(NioServerSocketChannel.class).childHandler(new ChannelInitializer<SocketChannel>() {
+          @Override
+          public void initChannel(SocketChannel ch) throws Exception {
+            if (initializer != null) {
+              initializer.accept(ch);
+            }
+            ch.pipeline().addLast("frameDecoder", new NettyFrameDecoder())
+                .addLast("frameEncoder", new NettyFrameEncoder()).addLast("handler", new NettyServerAvroHandler());
+          }
+        }).option(ChannelOption.SO_BACKLOG, 128).childOption(ChannelOption.SO_KEEPALIVE, true);
```
[the remark this answers is not in the document]
Comment (dkulp) [reply lang/java/ipc-netty/src/main/java/org/apache/avro/ipc/netty/NettyServer.java]: Done for both.

**Comment by dkulp** [comment]
> NettyServer -> new DefaultEventLoopGroup (16);
> I am not sure whether the number of threads 16 meets the needs of the job.
> In practice and application, regardless of memory consumption, I like to use the thread number formula to calculate
> Nthreads = Ncpu * Ucpu * (1 + w / c);
> So I think 16 can be used as the default value and the settings can be passed in through the constructor.

I updated NettyServer to allow passing EventLoopGroup objects for each of the three groups it uses.   Thus, the user could configure each of the three as needed.

**Comment by Fokko** [comment]
I was thinking about this, wouldn't it be better to shade netty?

**Pull request closed without merging**